core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -128,19 +133,17 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
     }
 
     def tryCreateControllerZNodeAndIncrementEpoch(): (Int, Int) = {
-      try {
-        val transaction = zooKeeperClient.createTransaction()
-        transaction.create(ControllerZNode.path, ControllerZNode.encode(controllerId, timestamp),
-          acls(ControllerZNode.path).asJava, CreateMode.EPHEMERAL)
-        transaction.setData(ControllerEpochZNode.path, ControllerEpochZNode.encode(newControllerEpoch), expectedControllerEpochZkVersion)
-        val results = transaction.commit()
-        val setDataResult = results.get(1).asInstanceOf[SetDataResult]
-        (newControllerEpoch, setDataResult.getStat.getVersion)
-      } catch {
-        case _: NodeExistsException | _: BadVersionException => checkControllerAndEpoch()
-        case _: ConnectionLossException =>
-          zooKeeperClient.waitUntilConnected()
-          tryCreateControllerZNodeAndIncrementEpoch()
+      val response = retryRequestUntilConnected(
```

Comment:
Now the patch is not longer using zookeeper transaction, will there be issue if e.g. controller znode is created but the controller epoch is not incremented?

Reply:
It is fine because we are using the zookeeper multi op directly right now. It is essentially the same as using zk transaction so we still provide the same guarantee.